File: src/server/vectorStore.ts
import fs from "fs";
import path from "path";
import { GoogleGenAI } from "@google/genai";
import { DocumentMeta, DocumentChunk, VectorSearchResult, VectorStoreStats } from "../types.js";

export const CATEGORIES = [
  "HR Policy",
  "Leave Policy",
  "IT Security Policy",
  "Code of Conduct",
  "Expense & Travel Policy",
  "Onboarding Guide",
];

const STORE_FILE_PATH = path.join(process.cwd(), "vector_store.json");
export const VECTOR_DIM = 768;

const STOP_WORDS = new Set([
  "a", "about", "above", "after", "again", "against", "all", "am", "an", "and", "any", "are", "aren't",
  "as", "at", "be", "because", "been", "before", "being", "below", "between", "both", "but", "by",
  "can", "can't", "cannot", "could", "couldn't", "did", "didn't", "do", "does", "doesn't", "doing",
  "don't", "down", "during", "each", "few", "for", "from", "further", "had", "hadn't", "has", "hasn't",
  "have", "haven't", "having", "he", "he'd", "he'll", "he's", "her", "here", "here's", "hers",
  "herself", "him", "himself", "his", "how", "how's", "i", "i'd", "i'll", "i'm", "i've", "if", "in",
  "into", "is", "isn't", "it", "it's", "its", "itself", "let's", "me", "more", "most", "mustn't", "my",
  "myself", "no", "nor", "not", "of", "off", "on", "once", "only", "or", "other", "ought", "our",
  "ours", "ourselves", "out", "over", "own", "same", "shan't", "she", "she'd", "she'll", "she's",
  "should", "shouldn't", "so", "some", "such", "than", "that", "that's", "the", "their", "theirs",
  "them", "themselves", "then", "there", "there's", "these", "they", "they'd", "they'll", "they're",
  "they've", "this", "those", "through", "to", "too", "under", "until", "up", "very", "was", "wasn't",
  "we", "we'd", "we'll", "we're", "we've", "were", "weren't", "what", "what's", "when", "when's",
  "where", "where's", "which", "while", "who", "who's", "whom", "why", "why's", "with", "won't",
  "would", "wouldn't", "you", "you'd", "you'll", "you're", "you've", "your", "yours", "yourself",
  "yourselves", "tell", "show", "give", "find", "search", "check", "please"
]);

export function cosineSimilarity(vecA: number[], vecB: number[]): number {
  if (!vecA || !vecB || vecA.length === 0 || vecA.length !== vecB.length) {
    return 0.0;
  }

  let dotProduct = 0.0;
  let normA = 0.0;
  let normB = 0.0;

  for (let i = 0; i < vecA.length; i++) {
    dotProduct += vecA[i] * vecB[i];
    normA += vecA[i] * vecA[i];
    normB += vecB[i] * vecB[i];
  }

  if (normA === 0 || normB === 0) {
    return 0.0;
  }
  return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
}

// Levenshtein distance for fuzzy typo correction (e.g. "hpurs" -> "hours")
function editDistance(a: string, b: string): number {
  if (a === b) return 0;
  if (a.length === 0) return b.length;
  if (b.length === 0) return a.length;

  const matrix: number[][] = [];
  for (let i = 0; i <= b.length; i++) matrix[i] = [i];
  for (let j = 0; j <= a.length; j++) matrix[0][j] = j;

  for (let i = 1; i <= b.length; i++) {
    for (let j = 1; j <= a.length; j++) {
      if (b.charAt(i - 1) === a.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }
  return matrix[b.length][a.length];
}

// Fuzzy matching score between two words (0.0 to 1.0)
function wordFuzzySimilarity(w1: string, w2: string): number {
  if (w1 === w2) return 1.0;
  if (w1.length < 3 || w2.length < 3) return w1 === w2 ? 1.0 : 0.0;
  
  // Prefix/contains check
  if (w1.length > 4 && w2.length > 4 && (w1.startsWith(w2) || w2.startsWith(w1))) {
    return 0.9;
  }

  const maxLen = Math.max(w1.length, w2.length);
  const dist = editDistance(w1, w2);
  const maxAllowedDist = maxLen <= 5 ? 1 : 2;

  if (dist <= maxAllowedDist) {
    return Math.max(0, 1.0 - dist / maxLen);
  }
  return 0.0;
}

export function computeLexicalAndFuzzyScore(query: string, chunkText: string, docName: string = ""): number {
  const cleanQ = query.toLowerCase().replace(/[^\w\s]/g, " ");
  const cleanText = (chunkText + " " + docName).toLowerCase().replace(/[^\w\s]/g, " ");

  const qWords = cleanQ.split(/\s+/).filter((w) => w.length > 1);
  const textWords = cleanText.split(/\s+/).filter((w) => w.length > 1);

  if (qWords.length === 0 || textWords.length === 0) return 0.0;

  // Filter query words to emphasize meaningful terms
  const meaningfulQWords = qWords.filter((w) => !STOP_WORDS.has(w));
  const targetQWords = meaningfulQWords.length > 0 ? meaningfulQWords : qWords;

  const textWordSet = new Set(textWords);
  let totalMatchWeight = 0.0;
  let maxPossibleWeight = 0.0;

  for (const qWord of targetQWords) {
    const wordWeight = STOP_WORDS.has(qWord) ? 0.3 : 1.0;
    maxPossibleWeight += wordWeight;

    // Exact match in text
    if (textWordSet.has(qWord)) {
      totalMatchWeight += wordWeight * 1.0;
      continue;
    }

    // Substring in full text
    if (cleanText.includes(qWord)) {
      totalMatchWeight += wordWeight * 0.95;
      continue;
    }

    // Fuzzy matching against all words in chunk
    let bestFuzzy = 0.0;
    for (const tWord of textWords) {
      if (Math.abs(tWord.length - qWord.length) <= 2) {
        const sim = wordFuzzySimilarity(qWord, tWord);
        if (sim > bestFuzzy) {
          bestFuzzy = sim;
          if (bestFuzzy >= 0.9) break;
        }
      }
    }

    if (bestFuzzy >= 0.6) {
      totalMatchWeight += wordWeight * bestFuzzy;
    }
  }

  let lexicalScore = maxPossibleWeight > 0 ? totalMatchWeight / maxPossibleWeight : 0.0;

  // Exact phrase boost (e.g. "working hours" appearing as continuous phrase in text)
  if (targetQWords.length >= 2) {
    const phrase = targetQWords.join(" ");
    if (cleanText.includes(phrase)) {
      lexicalScore = Math.min(1.0, lexicalScore * 1.35 + 0.2);
    }
  }

  return Math.min(1.0, lexicalScore);
}

export function generateFallbackEmbedding(text: string, dim: number = VECTOR_DIM): number[] {
  const cleaned = text.toLowerCase().replace(/[^\w\s]/g, " ");
  const words = cleaned.split(/\s+/).filter(Boolean);
  const vector = new Array(dim).fill(0.0);

  // Bag of words + Character n-grams (3-grams) for typo tolerance
  for (const word of words) {
    const isStop = STOP_WORDS.has(word);
    const weight = isStop ? 0.2 : 1.0;

    let hashVal = 0;
    for (let i = 0; i < word.length; i++) {
      hashVal = (Math.imul(hashVal, 31) + word.charCodeAt(i)) | 0;
    }
    const idx1 = Math.abs(hashVal) % dim;
    const idx2 = Math.abs((hashVal * 37) | 0) % dim;
    vector[idx1] += 1.0 * weight;
    vector[idx2] += 0.5 * weight;

    // Character 3-grams
    if (word.length >= 3) {
      for (let i = 0; i <= word.length - 3; i++) {
        const gram = word.slice(i, i + 3);
        let gHash = 0;
        for (let j = 0; j < gram.length; j++) {
          gHash = (Math.imul(gHash, 33) + gram.charCodeAt(j)) | 0;
        }
        const gIdx = Math.abs(gHash) % dim;
        vector[gIdx] += 0.3 * weight;
      }
    }
  }

  const norm = Math.sqrt(vector.reduce((acc, v) => acc + v * v, 0));
  if (norm > 0) {
    return vector.map((v) => v / norm);
  }

  return vector;
}

export async function getEmbedding(text: string, apiKey?: string): Promise<number[]> {
  const key = apiKey || process.env.GEMINI_API_KEY;
  if (!key) {
    return generateFallbackEmbedding(text);
  }

  const ai = new GoogleGenAI({ apiKey: key });

  const models = ["gemini-embedding-001", "gemini-embedding-2", "gemini-embedding-2-preview"];
  for (const model of models) {
    try {
      const response = await ai.models.embedContent({
        model,
        contents: text,
      });
      const values = (response as any).embedding?.values || (response as any).embeddings?.[0]?.values;
      if (values && values.length > 0) {
        return Array.from(values);
      }
    } catch {
      continue;
    }
  }

  return generateFallbackEmbedding(text);
}

export class VectorDatabase {
  private documents: Map<string, DocumentMeta> = new Map();
  private chunks: Map<string, DocumentChunk> = new Map();

  constructor() {
    this.loadFromDisk();
  }

  private saveToDisk(): void {
    try {
      const data = {
        documents: Array.from(this.documents.values()),
        chunks: Array.from(this.chunks.values()),
      };
      fs.writeFileSync(STORE_FILE_PATH, JSON.stringify(data, null, 2), "utf-8");
    } catch (e) {
      console.warn(`Failed to persist vector store to disk: ${e}`);
    }
  }

  private loadFromDisk(): void {
    try {
      if (fs.existsSync(STORE_FILE_PATH)) {
        const raw = fs.readFileSync(STORE_FILE_PATH, "utf-8");
        if (raw.trim()) {
          const parsed = JSON.parse(raw);
          for (const doc of parsed.documents || []) {
            this.documents.set(doc.id, doc);
          }
          for (const chunk of parsed.chunks || []) {
            this.chunks.set(chunk.id, chunk);
          }
        }
      }
    } catch (e) {
      console.warn(`Failed to load vector store from disk: ${e}`);
    }
  }

  public getDocuments(): DocumentMeta[] {
    return Array.from(this.documents.values());
  }

  public getDocument(docId: string): DocumentMeta | undefined {
    return this.documents.get(docId);
  }

  public getChunksForDoc(docId: string): DocumentChunk[] {
    return Array.from(this.chunks.values()).filter((c) => c.docId === docId);
  }

  public getAllChunks(): DocumentChunk[] {
    return Array.from(this.chunks.values());
  }

  public addDocument(doc: DocumentMeta, docChunks: DocumentChunk[]): void {
    this.documents.set(doc.id, doc);
    for (const chunk of docChunks) {
      this.chunks.set(chunk.id, chunk);
    }
    this.saveToDisk();
  }

  public updateChunkEmbedding(chunkId: string, embedding: number[]): void {
    const chunk = this.chunks.get(chunkId);
    if (chunk) {
      chunk.embedding = embedding;
      this.saveToDisk();
    }
  }

  public deleteDocument(docId: string): boolean {
    if (this.documents.has(docId)) {
      this.documents.delete(docId);
      for (const [cid, c] of this.chunks.entries()) {
        if (c.docId === docId) {
          this.chunks.delete(cid);
        }
      }
      this.saveToDisk();
      return true;
    }
    return false;
  }

  public clearAll(): void {
    this.documents.clear();
    this.chunks.clear();
    this.saveToDisk();
  }

  public getCategories(): string[] {
    const cats = new Set<string>();
    for (const doc of this.documents.values()) {
      cats.add(doc.category || "General");
    }
    return Array.from(cats).sort();
  }

  /**
   * Hybrid Search: Combines Dense Neural Vector Similarity + Sparse Lexical & Fuzzy Typo Matching.
   * This guarantees that exact & near-match queries (e.g. "Working hpurs", "working hours", "leave rules")
   * as well as conceptual/semantic queries retrieve the most accurate document chunks.
   */
  public async search(
    queryEmbedding: number[],
    queryString: string = "",
    topK: number = 4,
    threshold: number = 0.1,
    selectedDocIds?: string[] | null,
    selectedCategories?: string[] | null
  ): Promise<VectorSearchResult[]> {
    const results: VectorSearchResult[] = [];

    for (const chunk of this.chunks.values()) {
      if (selectedDocIds && selectedDocIds.length > 0 && !selectedDocIds.includes(chunk.docId)) {
        continue;
      }

      if (selectedCategories && selectedCategories.length > 0) {
        const doc = this.documents.get(chunk.docId);
        if (!doc || !selectedCategories.includes(doc.category || "General")) {
          continue;
        }
      }

      const vectorSim = cosineSimilarity(queryEmbedding, chunk.embedding);
      const lexicalFuzzySim = queryString
        ? computeLexicalAndFuzzyScore(queryString, chunk.text, chunk.docName)
        : 0.0;

      // Hybrid score:
      // If neural vector similarity is strong (>0.25), combine with lexical weight.
      // If neural similarity is 0 (e.g., embedding model mismatch or fallback), lexical/fuzzy dominates.
      let finalScore = 0.0;
      if (vectorSim > 0.1) {
        finalScore = vectorSim * 0.6 + lexicalFuzzySim * 0.4;
      } else if (lexicalFuzzySim > 0.0) {
        finalScore = lexicalFuzzySim * 0.95;
      } else {
        finalScore = vectorSim;
      }

      if (finalScore >= threshold) {
        results.push({
          chunk,
          similarity: Number(finalScore.toFixed(4)),
        });
      }
    }

    results.sort((a, b) => b.similarity - a.similarity);
    return results.slice(0, topK);
  }

  public getStats(): VectorStoreStats {
    const allChunks = Array.from(this.chunks.values());
    const totalChars = allChunks.reduce((acc, c) => acc + (c.charLength || 0), 0);
    const avgChunkLength = allChunks.length > 0 ? Math.round(totalChars / allChunks.length) : 0;
    const dimensions =
      allChunks.length > 0 && allChunks[0].embedding?.length
        ? allChunks[0].embedding.length
        : VECTOR_DIM;

    let storageSizeBytes = 0;
    try {
      if (fs.existsSync(STORE_FILE_PATH)) {
        storageSizeBytes = fs.statSync(STORE_FILE_PATH).size;
      }
    } catch {
      storageSizeBytes = JSON.stringify(allChunks).length;
    }

    return {
      totalDocs: this.documents.size,
      totalChunks: this.chunks.size,
      vectorDimensions: dimensions,
      storageSizeBytes,
      avgChunkLength,
    };
  }
}

export const globalVectorDb = new VectorDatabase();

File: backend/vector_store.py
"""In-memory vector database with JSON persistence (ported from src/lib/vectorStore.ts)."""

import json
import math
import os
import re
from pathlib import Path
from typing import Optional

from google import genai

STORE_FILE_PATH = Path(__file__).resolve().parent.parent / "vector_store.json"
VECTOR_DIM = 768  # standard embedding size for text-embedding-004


def cosine_similarity(vec_a: list[float], vec_b: list[float]) -> float:
    """Cosine similarity between two vectors."""
    if not vec_a or not vec_b or len(vec_a) == 0 or len(vec_a) != len(vec_b):
        return 0.0

    dot_product = 0.0
    norm_a = 0.0
    norm_b = 0.0

    for a, b in zip(vec_a, vec_b):
        dot_product += a * b
        norm_a += a * a
        norm_b += b * b

    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot_product / (math.sqrt(norm_a) * math.sqrt(norm_b))


def _to_int32(value: int) -> int:
    """Emulates JavaScript's `value | 0` 32-bit signed integer overflow."""
    value &= 0xFFFFFFFF
    if value >= 0x80000000:
        value -= 0x100000000
    return value


def generate_fallback_embedding(text: str, dim: int = VECTOR_DIM) -> list[float]:
    """Deterministic fallback vector generator, mirrors the JS hashing vectorizer 1:1."""
    cleaned = re.sub(r"[^\w\s]", "", text.lower())
    words = [w for w in re.split(r"\s+", cleaned) if w]
    vector = [0.0] * dim

    for word in words:
        hash_val = 0
        for ch in word:
            shifted = _to_int32(hash_val << 5)
            hash_val = _to_int32(shifted - hash_val + ord(ch))

        primary_index = abs(hash_val) % dim
        secondary_index = abs(hash_val * 31) % dim

        vector[primary_index] += 1.0
        vector[secondary_index] += 0.5

    norm = math.sqrt(sum(v * v for v in vector))
    if norm > 0:
        vector = [v / norm for v in vector]

    return vector


async def get_embedding(text: str, api_key: Optional[str] = None) -> list[float]:
    """Generates vector embeddings for a given string using Gemini API or fallback."""
    key = api_key or os.environ.get("GEMINI_API_KEY")
    if not key:
        return generate_fallback_embedding(text)

    models_to_try = ["gemini-embedding-2-preview", "text-embedding-004"]
    client = genai.Client(api_key=key)

    for model in models_to_try:
        try:
            response = await client.aio.models.embed_content(model=model, contents=text)
            values = None
            if response.embeddings and len(response.embeddings) > 0:
                values = response.embeddings[0].values
            if values:
                return list(values)
        except Exception:
            continue

    return generate_fallback_embedding(text)


class VectorDatabase:
    def __init__(self) -> None:
        self.documents: dict[str, dict] = {}
        self.chunks: dict[str, dict] = {}
        self._load_from_disk()

    def _save_to_disk(self) -> None:
        try:
            data = {
                "documents": list(self.documents.values()),
                "chunks": list(self.chunks.values()),
            }
            STORE_FILE_PATH.write_text(json.dumps(data, indent=2), encoding="utf-8")
        except Exception as e:
            print(f"Failed to persist vector store to disk: {e}")

    def _load_from_disk(self) -> None:
        try:
            if STORE_FILE_PATH.exists():
                parsed = json.loads(STORE_FILE_PATH.read_text(encoding="utf-8"))
                for doc in parsed.get("documents", []):
                    self.documents[doc["id"]] = doc
                for chunk in parsed.get("chunks", []):
                    self.chunks[chunk["id"]] = chunk
        except Exception as e:
            print(f"Failed to load vector store from disk: {e}")

    def get_documents(self) -> list[dict]:
        return list(self.documents.values())

    def get_document(self, doc_id: str) -> Optional[dict]:
        return self.documents.get(doc_id)

    def get_chunks_for_doc(self, doc_id: str) -> list[dict]:
        return [c for c in self.chunks.values() if c["docId"] == doc_id]

    def get_all_chunks(self) -> list[dict]:
        return list(self.chunks.values())

    def add_document(self, doc: dict, doc_chunks: list[dict]) -> None:
        self.documents[doc["id"]] = doc
        for chunk in doc_chunks:
            self.chunks[chunk["id"]] = chunk
        self._save_to_disk()

    def delete_document(self, doc_id: str) -> bool:
        if doc_id in self.documents:
            del self.documents[doc_id]
            for chunk_id in [cid for cid, c in self.chunks.items() if c["docId"] == doc_id]:
                del self.chunks[chunk_id]
            self._save_to_disk()
            return True
        return False

    def clear_all(self) -> None:
        self.documents.clear()
        self.chunks.clear()
        self._save_to_disk()

    def get_categories(self) -> list[str]:
        return sorted({doc.get("category", "General") for doc in self.documents.values()})

    async def search(
        self,
        query_embedding: list[float],
        top_k: int = 4,
        threshold: float = 0.1,
        selected_doc_ids: Optional[list[str]] = None,
        selected_categories: Optional[list[str]] = None,
    ) -> list[dict]:
        results: list[dict] = []

        for chunk in self.chunks.values():
            if selected_doc_ids and len(selected_doc_ids) > 0 and chunk["docId"] not in selected_doc_ids:
                continue

            if selected_categories and len(selected_categories) > 0:
                doc = self.documents.get(chunk["docId"])
                if not doc or doc.get("category", "General") not in selected_categories:
                    continue

            similarity = cosine_similarity(query_embedding, chunk["embedding"])
            if similarity >= threshold:
                results.append({"chunk": chunk, "similarity": similarity})

        results.sort(key=lambda r: r["similarity"], reverse=True)
        return results[:top_k]

    def get_stats(self) -> dict:
        all_chunks = list(self.chunks.values())
        total_chars = sum(c["charLength"] for c in all_chunks)
        avg_chunk_length = round(total_chars / len(all_chunks)) if all_chunks else 0
        dimensions = len(all_chunks[0]["embedding"]) if all_chunks and all_chunks[0].get("embedding") else VECTOR_DIM

        storage_size_bytes = 0
        try:
            if STORE_FILE_PATH.exists():
                storage_size_bytes = STORE_FILE_PATH.stat().st_size
        except Exception:
            storage_size_bytes = len(json.dumps(all_chunks))

        return {
            "totalDocs": len(self.documents),
            "totalChunks": len(self.chunks),
            "vectorDimensions": dimensions,
            "storageSizeBytes": storage_size_bytes,
            "avgChunkLength": avg_chunk_length,
        }


global_vector_db = VectorDatabase()
